dvc/annotations.py

```diff
@@ -26,10 +26,20 @@ def to_dict(self) -> Dict[str, str]:
         return compact(asdict(self))
 
 
+@dataclass
+class Artifact(Annotation):
+    PARAM_PATH: ClassVar[str] = "path"
+    path: Optional[str] = None
```

Comment:
Ok, so we discussed this - and I implemented this on Friday: if `path` isn't specified, we use `ID` instead of path.

But after discussing this with @omesser again, looks like asking for `path` is preferable, since names in Model Registry won't be pretty with arbitrary `path`s used as `name`s .

So, I made another commit here, that actually requires entering `path`, and sets more readable limitations on `name`s.

Beside that, I disallowed using `/` in artifact name, since this will be easier now for monorepo case (I don't remember users using this, although the reason was to allow something like "namespaces"). E.g. for DVClive, whatever the `path` to artifact will be, Git tag will be named `dvclive/my-artifact` - and we don't need to figure out what to do with diallowed `:` in Git tags like `dvclive:my-artifact`. One option here was to replace `:` with `/`, but then we can potentially get ambigious results - I do this now, and we can lift this later (and get back to the question of "namespaces" - or let's wait for user feedback here).

My premise for all these cases is: let's set stronger limitations (`path` is required, using `/` is disallowed), cause it's easy to lift them later if we wish. Add them back is harder.

With that said, let's merge this. I don't have rights to merge - so either please give me them, or do it yourself. cc @dberenbaum @daavoo @omesser

Reply:
I still have one concern here: will `live.log_artifact()` require both a path and a name argument? Can we infer a name if one is not provided? 

Asking for both a path and name may feel odd to people coming from other tools like mlflow, and it's not immediately clear why they need to do this or what benefit they get. It no longer feels like they get the model registry "for free."

Reply:
> Can we infer a name if one is not provided?

I don't want to block the P.R, but, given the last agreement on making `path` mandatory, why don't we make `path` the primary key and `name` just another optional field?

Reply:
> But after discussing this with @omesser again, looks like asking for `path` is preferable, since names in Model Registry won't be pretty with arbitrary `path`s used as `name`s .

AFAIK the point from @aguschin and @omesser above is to require a human-readable name that is not a path so it looks better in the model registy UI.

Reply:
> AFAIK the point from @aguschin and @omesser above is to require a human-readable name that is not a path so it looks better in the model registy UI.

TBH, I don't fully agree on requiring the users to provide something that is not strictly necessary, only because we _think_ that it looks better (and it's what users want) on the UI. We could always *prefer* `name` on the UI without making it mandatory.

Again, I am ok with the current status, just wanted to fully discuss/understand the reasoning for going with `name` as the primary key as it would be very hard to change once we merge&release.

Reply:
I think it's worth blocking to come to some agreement on this.

Reply:
@daavoo WDYT of the idea to infer a name in DVCLive (like use the basename of the file and drop the extension)? Since DVCLive would be the primary workflow we will document for logging artifacts, this would go a long way towards keeping clean names without requiring users to name each artifact.

Reply:
in cont. to @aguschin 's comments, adding my thoughts and some context:

### needing a name (which is not a path)
I understand this might be a point of contention (inconsistent compared to current DVC yaml structure), but I believe some key/display-name which is NOT the full path is necessary for our registry feature-set / use case.
- releases, stages - path is not compatible with GTO - and we must support those actions
- using arbitrary path **as a display name** is basically a no-go in a GUI - I'm not saying we can't infer it, but we can't just ignore the need for a display name which is not a path. Using only paths in a registry GUI can make it down right unusable IMO - no competitor is doing it as well afaik. so when thinking of the full experience - names are definitely needed. I also thinks it makes more sense as a name to refer to the model/artifact if we'll need to refer to it in other places in dvc.yaml - more friendly than using the full path.

### just infer the name (and make it optional)
IMO that can work fine actually - something like this: `{normalized-basename-without-ext}-{first-4-chars-of-md5(path)}` or something similar?
I expect in most registry-scenarios users would just provide name explicitly (also suggest to specify it in docs) - but if someone thinks specifying a name for your model is friction - 👍 
-> So, in live API `name` can be optional

### dvc.yaml artifacts structure
I would strongly suggest we pick a schema and stick with it, and not try to mix (so the artifact key has dual meaning and validation based on some field existing or not).

In this case, 2 options are either:
1. path is key, `name` optional:
   ```yaml
   artifacts:
     ./some/path/model.ext:
       type: model
       name: my-model. # optional
   ```

2. name is key, `path` mandatory:
   ```yaml
   artifacts:
     my-model:
       type: model
       path: ./some/path/model.ext  # mandatory
   ```

Here, while (2) looks more natural to me out-of-context (file path as key? unnatural + potentially fragile 😄 ) - I don't have any strong feelings either way.
(1) Seems more agreeable with existing approach of path as keys, and name is optional seems important to some.

@aguschin  - will (1) be problematic / no go?  wdyt?

Reply:
These can be considered as corner cases that's out of happy path to some extent I guess. But the edits I need to do in `dvc.yaml` don't fit my "mental model" of MR, and I think that won't be the case for me only.

Reply:
> ^ artifact 2 don't have name, thus using path. Ok, this is what's expected, so no big questions here.

In my opinion the tags will ONLY have the artifact name. even if it's "inferred" (e.g. `model2-d35h`). 
I agree that having git tags / UI display names (all user facing) referring to names not explicitly appearing / mandates in dvc.yaml is a bit janky though (another reason I preferred option (2) above).
For what it's worth - I second the opinion it is more natural for the registry use case, even if less conforming with existing DVC-ness

Reply:
I agree with names as top-level keys (I only thought we should allow them to double as paths), and @daavoo has said he doesn't intend to block the PR. It sounds like we are in agreement, so let's merge.

As a follow-up, why does GTO require name and path as distinct fields? It seems like there are some scenarios where the path is short and meaningful. In those scenarios, I could use all the GTO commands without any `dvc.yaml` artifacts through the git tags+paths.

Also, a minor opinion: I'm not sure about using the tag format `dvclive/my-artifact` since it will look like a path and could be confusing, so I'd prefer we use `+` or some other separator.

Reply:
> why does GTO require name and path as distinct fields? It seems like there are some scenarios where the path is short and meaningful.

cc @aguschin to answer that from original GTO of view.
in my opinion - the ambiguity of the fields that will be created is not worth trying to "save a field" for the user. clarity/consistency over brevity.